django/utils/hashable.py

```diff
@@ -0,0 +1,11 @@
+from collections import Iterable
+
+
+def make_hashable(value):
+    if isinstance(value, dict):
+        return tuple([
+            (key, make_hashable(nested_value)) for key, nested_value in value.items()
+        ])
+    if isinstance(value, Iterable) and not isinstance(value, str):
```

Comment:
Another issue with checking for `isinstance(value, Iterable)` unconditionally is that the object could be iterable and implement `__hash__` and we wouldn't respect it.

Maybe attempting to `hash` the value and only falling back on `TypeError` is more appropriate?

```python
if isinstance(value, dict):
    return tuple([
        (key, make_hashable(nested_value)) for key, nested_value in value.items()
    ])
try:
    hash(value)
except TypeError:
    if is_iterable(value):
        return tuple(map(make_hashable, value))
    raise
return value

Reply:
Simon, is the `is_iterable()` check to avoid hiding a bug in a `__hash__` that raises `TypeError`? I'm trying to add a test for that.

Reply:
So things like list and set are going to going to hit the try, they won't be hashable, and then they'll hit the check, `is_iterable`. Things like strings are going to hit try hash() and then go through, since string has `__hash__` implemented. 

I think the `raise` will be hit when we get some less simple data structure in there that isn't iterable and doesn't have `__hash__` implemented.